File: leet_lint_code/leet_code/subsets.cpp
// Copyright © 2018 Hailin Gu. All rights reserved.
// License(GPL)
// Author: Hailin Gu
// This is a answer of leet code problem 78.

#ifndef LEETLINTCODE_LINTCODE_SUBSETS_CPP
#define LEETLINTCODE_LINTCODE_SUBSETS_CPP

#include <iostream>
#include <vector>
#include <algorithm>

class Subsets {
 public:
   std::vector<std::vector<int> > subsets(std::vector<int> &nums) {
     std::vector<std::vector<int> > ret;
     sort(nums.begin(), nums.end());
     if (0 == nums.size()) {
      std::vector<int> v;
      ret.push_back(v);
    } else subsetsHelp(nums, ret);
    return ret;
   }

   void subsetsHelp(std::vector<int>& s, std::vector<std::vector<int> >& save){
    int p;
    if (s.size() > 1){
        p = s[s.size() - 1];
        s.pop_back();
        subsetsHelp(s, save);
        int t_size = (int)save.size();
        for (int i = 0; i < t_size; ++ i) {
            std::vector<int> temp = save[i];
            temp.push_back(p);
            save.push_back(temp);
        }
        return ;
    }

    p = s[0];
    std::vector<int> temp;
    std::vector<int> empty;
    temp.push_back(p);
    save.push_back(empty);
    save.push_back(temp);
  }
};

#endif // LEETLINTCODE_LINTCODE_SUBSETS_CPP

File: leet_lint_code/leet_code/java/Subsets.java
// Copyright © 2018 Hailin Gu. All rights reserved.
// License
// Author: Hailin Gu
// This is a answer of leet code problem 78.
// Date: 2019.7.29

import java.util.List;
import java.util.ArrayList;
import java.util.Arrays;


class Subsets {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> ans = new ArrayList<>();
        Arrays.sort(nums);
        
        List<List<Integer>> tmp = new ArrayList<>();
        for (int i : nums) {
             tmp.clear();
            for (List<Integer> list: ans) {
                List<Integer> nList = new ArrayList<Integer>(list);
                nList.add(i);
                tmp.add(nList);
            }

            List<Integer> n = new ArrayList<>();
            n.add(i);
            tmp.add(n);
            ans.addAll(tmp);
        }

        ans.add(new ArrayList<>());
        return ans;
    }
}